### PR TITLE
Create LCS.c

### PR DESCRIPTION
**Pull request by Santoshjonnakuti**
Added LCS algorithm

#### Description of Change
Added Longest common subsequence algorithm

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#File-Name-guidelines)
- [ ] Added tests and example, test must pass
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: feat: Add LCS Algorithm

### CODE REVIEW

**Comment by Santoshjonnakuti** [comment]
Please review and let me know if something is wrong @Panquesito7
Thank you!

**Comment by Panquesito7** [comment]
> Please review and let me know if something is wrong @Panquesito7

I'll review it as soon as I can. Join our [Discord server](https://discord.gg/c7MnfGFGa6) to stay tuned. Thanks. 🙂

File: dynamic_programming/LCS.c [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+ */
+
+
+#include<stdio.h>     /// for I/O operations
+#include<stdlib.h>   ///  for Dynamic memory allocation
+#include<string.h>  ///   for puts(), gets() and strlen()
+
+#define MAX 100      // Declaration of a constant value
+
+int lengthOfLCS;
```
Comment (Panquesito7) [inline dynamic_programming/LCS.c:35]: Avoid using global variables. Pass them as function parameters or as struct members instead.

File: dynamic_programming/LCS.c [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+/**
+ * @brief prints all possible Longest common subsequences between string1 and string2 using recursion
+ * @param c 2D-Array of Optimal solutions of all sub-problems
+ * @param a Longest common subsequence 
+ * @param index  length of a
+ * @param i length of string 1
+ * @param j length of string 2
+ * @param string1 String 1
+ * @param string2 String 2
+ * @prints one of te Longest Common Subsequence
```
Comment (Panquesito7) [inline dynamic_programming/LCS.c:52]:
```suggestion
 * @prints one of the Longest Common Subsequence
```

File: dynamic_programming/LCS.c [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+			}
+			else
+			{
+				j--;
+			}
+		}
+	}
+	return a;
+}
+void print(char *a, int index)
```
Comment (Panquesito7) [inline dynamic_programming/LCS.c:89]: Missing function documentation.

File: dynamic_programming/LCS.c [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+	return b;
+}
+
+/**
+ * @breif LCS will make 2D-array with optimal solutions of sub-problems
+ * @param string1 String 1
+ * @param m Length of String 1
+ * @param string2 String 2
+ * @param n Length of String 2
+ * returns 2D-array.
```
Comment (Panquesito7) [inline dynamic_programming/LCS.c:121]:
```suggestion
 * @returns 2D-array
```

File: dynamic_programming/LCS.c [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+	}
+	printf("\n\n");
+	return;
+}
+
+/**
+ * @breif max will find the maximum between 2 values
+ * @param a integer a
+ * @param b integer b
+ * returns maximum value
```
Comment (Panquesito7) [inline dynamic_programming/LCS.c:104]:
```suggestion
 * @returns maximum value
```

File: dynamic_programming/LCS.c (hunk 2 above)
Comment (Panquesito7) [inline dynamic_programming/LCS.c:52]: What does this function return? Every function returns something unless they are `void`. Please document it here.

File: dynamic_programming/LCS.c [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+{
+	if(a > b)
+	{
+		return a;
+	}
+	return b;
+}
+
+/**
+ * @breif LCS will make 2D-array with optimal solutions of sub-problems
```
Comment (Panquesito7) [inline dynamic_programming/LCS.c:116]:
```suggestion
 * @brief LCS will make 2D-array with optimal solutions of sub-problems
```

File: dynamic_programming/LCS.c [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+ *
+ * So There are 3 Longest common subsequences in 'GAC' and 'AGCAT'
+ */
+
+
+#include<stdio.h>     /// for I/O operations
+#include<stdlib.h>   ///  for Dynamic memory allocation
+#include<string.h>  ///   for puts(), gets() and strlen()
+
+#define MAX 100      // Declaration of a constant value
```
Comment (Panquesito7) [inline dynamic_programming/LCS.c:33]: Provide a better description of this constant variable.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
👍 Good work!
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/Santoshjonnakuti/C/actions

Please rename the file to `longest_common_subsequence.cpp`.

File: dynamic_programming/LCS.c (hunk 7 above)
Comment (Santoshjonnakuti) [reply dynamic_programming/LCS.c]: Updated all the Changes Requested Please review @Panquesito7  as soon as you can.
Thank You!

File: dynamic_programming/longest_common_subsequence.cpp [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+	/// Assigning Values of 'AGCAT' for string2
+	string2[0] = 'A';
+	string2[1] = 'G';
+	string2[2] = 'C';
+	string2[3] = 'A';
+	string2[4] = 'T';
+	
+	// Lengths of two Strings;
+	int lenOfStr1 = strlen(string1);
+	int lenOfStr2 = strlen(string2);
```
Comment (Panquesito7) [inline dynamic_programming/longest_common_subsequence.cpp:185]: Add a one-line description of the variables (see the line 167, 168, and 169).

**Review by Panquesito7 — COMMENTED** [review]
👍 Good work!

File: dynamic_programming/longest_common_subsequence.cpp (hunk 8 above)
Comment (Santoshjonnakuti) [reply dynamic_programming/longest_common_subsequence.cpp]: Updated all the Changes Requested Please review @Panquesito7 as soon as you can.
Thank You!

File: dynamic_programming/longest_common_subsequence.cpp [hunk 9]
```diff
@@ -0,0 +1,208 @@
+/**
+ * @file
+ * @author [Santosh Jonnakuti](https://github.com/Santoshjonnakuti)
+ * @brief [Longest Common Subsequence Algorithm](https://en.wikipedia.org/wiki/Longest_common_subsequence_problem)
+ * implementation in C to find the LCS between 2 strings.
+ * @details LCS algorithm uses dynamic approach to find all possible LCS between 2 strings.
```
Comment (ayaankhan98) [inline dynamic_programming/longest_common_subsequence.cpp:6]:
```suggestion
 * @details LCS algorithm uses dynamic programming approach to find all possible LCS between 2 strings. 
```

File: dynamic_programming/longest_common_subsequence.cpp [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+    for(int i=index-1; i>=0; i--)
+	{
+		printf("%c", a[i]);
+	}
+	printf("\n\n");
+	return;
+}
+
+/**
+ * @breif max will find the maximum between 2 values
```
Comment (ayaankhan98) [inline dynamic_programming/longest_common_subsequence.cpp:95]:
```suggestion
 * @brief max will find the maximum between 2 values
```

File: dynamic_programming/longest_common_subsequence.cpp [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+	for(int i=1; i<=m; i++)
+	{
+		c[i][0] = 0;
+	}
+	
+	// If length of string 1 is 0 then there will be no Longest common subsequence
+	for(int j=0; j <=n; j++)
+	{
+		c[0][j] = 0;
+	}
```
Comment (ayaankhan98) [inline dynamic_programming/longest_common_subsequence.cpp:142]: In these two cases, this function should directly return to the procedure which called it without doing any further processing.

File: dynamic_programming/longest_common_subsequence.cpp [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+	
+	// If there is no Longest common subsequence
+	if (!lengthOfLCS)
+	{
+		printf("There is no Possible LCS...\n\n");
+		return 0;
+	}
+	// If there is a Longest common subsequence then printing them
+	printf("All Possible LCS are : \n\n");
+	allPossibleLCS(c, a, index, lenOfStr1, lenOfStr2, string1, string2);
```
Comment (ayaankhan98) [inline dynamic_programming/longest_common_subsequence.cpp:206]: please write a separate test function with some tests (including corner cases) to verify the implementation.

**Review by ayaankhan98 — COMMENTED** [review]
Thanks for contribution. please consider above suggestions.

File: dynamic_programming/longest_common_subsequence.cpp (hunk 12 above)
Comment (ayaankhan98) [reply dynamic_programming/longest_common_subsequence.cpp:206]: Also don't forget to release the dynamically allocated memory resources.

Comment (Santoshjonnakuti) [reply dynamic_programming/longest_common_subsequence.cpp]: Can I use some global variables to write the test function?

Comment (ayaankhan98) [reply dynamic_programming/longest_common_subsequence.cpp]: It's not a good practice to use global variables.

Comment (Santoshjonnakuti) [reply dynamic_programming/longest_common_subsequence.cpp]: Thanks for the suggestion @ayaankhan98 !
I'll try my best.

File: dynamic_programming/longest_common_subsequence.cpp [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+ * GA
+ * GA
+ * 
+ * So There are 3 Longest common subsequences in 'GAC' and 'AGCAT'
+ */
+
+
+#include<stdio.h>     /// for I/O operations
+#include<stdlib.h>   ///  for Dynamic memory allocation
+#include<string.h>  ///   for puts(), gets() and strlen()
```
Comment (Panquesito7) [inline dynamic_programming/longest_common_subsequence.cpp:32]:
```suggestion
#include <stdio.h>     /// for I/O operations
#include <stdlib.h>   ///  for dynamic memory allocation
#include <string.h>  ///   for `puts()`, `gets()`, and `strlen()`
```

File: dynamic_programming/longest_common_subsequence.cpp [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+
+#define MAX 100      // maximum size of string1 and string2
+
+int lengthOfLCS;      // to store length of LCS 
+char **LCSArray;      // 2D- Character array to store all possible LCS
+int ind=0;
```
Comment (Panquesito7) [inline dynamic_programming/longest_common_subsequence.cpp:38]: Avoid using global variables as they may cause overflows and other issues. Pass them as local function parameters or as an structure members.

File: dynamic_programming/longest_common_subsequence.cpp [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+		}
+	}
+	return c;
+}
+
+/**
+ * @brief test is to test the algorithm
+ * @returns void
+ */
+void findAllLCS(char *string1, char *string2)
```
Comment (Panquesito7) [inline dynamic_programming/longest_common_subsequence.cpp:189]: The test function should use `assert` and be named `test` (see the [typical structure](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) of a program).

Comment (Santoshjonnakuti) [reply dynamic_programming/longest_common_subsequence.cpp:198]: Added test function with assert

File: dynamic_programming/longest_common_subsequence.cpp [hunk 16]
```diff
@@ -0,0 +1,301 @@
...
+ * 
+ * To test run './longest_common_subsequence -test'
+ * So There are 4 Longest common subsequences in 'GAC' and 'AGCAT'
+ */
+
+
+#include<stdio.h>      // for I/O operations
+#include<stdlib.h>     // for dynamic memory allocation
+#include<string.h>     // for `puts()`, `gets()`, and `strlen()`
+#include<assert.h>     // for `assert()`
```
Comment (Panquesito7) [inline dynamic_programming/longest_common_subsequence.cpp:34]:
```suggestion
#include <stdio.h>      /// for I/O operations
#include <stdlib.h>     /// for dynamic memory allocation
#include <string.h>     /// for `puts()`, `gets()`, and `strlen()`
#include <assert.h>     /// for `assert()`
```

File: dynamic_programming/longest_common_subsequence.cpp [hunk 17]
```diff
@@ -0,0 +1,301 @@
...
+
+#define MAX 100      // maximum size of string1 and string2
+
+// structure to store all LCS and no of possible LCS
```
Comment (Panquesito7) [inline dynamic_programming/longest_common_subsequence.cpp:38]:
```suggestion
/**
 * @brief Structure to store all LCS and no of possible LCS
 */
```

File: dynamic_programming/longest_common_subsequence.cpp [hunk 18]
```diff
@@ -0,0 +1,301 @@
...
+
+    // To run the test
+    if(argc == 2 && strcmp(argv[1], "-test") == 0)
+    {
+        test();
+        return 0;
+    }
+
+    // User input case
+	printf("Enter string1 : ");
```
Comment (Panquesito7) [inline dynamic_programming/longest_common_subsequence.cpp:295]: Wrong indentation.

File: dynamic_programming/longest_common_subsequence.cpp [hunk 19]
```diff
@@ -0,0 +1,301 @@
...
+    LCSArray = NULL;
+    free(LCSArray);
+    return;
+}
+
+/**
+ * @brief test is to test the algorithm
+ * @returns void
+ */
+void test()
```
Comment (Panquesito7) [inline dynamic_programming/longest_common_subsequence.cpp:254]:
```suggestion
static void test()
```

File: dynamic_programming/longest_common_subsequence.cpp [hunk 20]
```diff
@@ -0,0 +1,301 @@
...
+	free(string2);
+    LCSArray = NULL;
+    free(LCSArray);
+    return;
+}
+
+/**
+ * @brief test is to test the algorithm
+ * @returns void
+ */
```
Comment (Panquesito7) [inline dynamic_programming/longest_common_subsequence.cpp:253]:
```suggestion
/**
 * @brief Self-test implementations
 * @returns void
 */
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
👍 Good work.

File: dynamic_programming/longest_common_subsequence.cpp (hunk 20 above)
Comment (Santoshjonnakuti) [reply dynamic_programming/longest_common_subsequence.cpp]: Updated!

File: dynamic_programming/longest_common_subsequence.cpp [hunk 21]
```diff
@@ -0,0 +1,302 @@
...
+ * @param argc commandline argument count (ignored)
+ * @param argv commandline array of arguments (ignored)
+ * @returns 0 on exit
+ */
+int main(int argc, char const *argv[])
+{
+    /// string1 and string2 are used to store strings in which we have to find LCS
+	char *string1, *string2;                      ///< 2 strings of max size MAX 
+	string1 = (char *)calloc(MAX, sizeof(char));  ///< memory allocation for string 1
+	string2 = (char *)calloc(MAX, sizeof(char));  ///< memory allocation for string 2
```
Comment (ayaankhan98) [inline dynamic_programming/longest_common_subsequence.cpp:287]: where this memory gets deallocated?

Comment (Santoshjonnakuti) [reply dynamic_programming/longest_common_subsequence.cpp:287]: Sorry! for that
Fixed Now

**Comment by Santoshjonnakuti** [comment]
@Panquesito7  Please review this as soon as you can
Thank you!

File: dynamic_programming/longest_common_subsequence.cpp [hunk 22]
```diff
@@ -0,0 +1,304 @@
...
+    int length;           // no of possible LCs
+} variable;
+
+
+// Functions Declaration
+int max(int a, int b);
+int** LCS(char *string1, int m, char *string2, int n);
+void allPossibleLCS(int **c, char *a, int index, int i, int j, char *string1, char *string2, int lengthOfLCS);
+void findAllLCS(char *string1, char *string2);
+void test();
```
Comment (Panquesito7) [inline dynamic_programming/longest_common_subsequence.cpp:53]:
```suggestion
static void test();
```

**Review by Panquesito7 — COMMENTED** [review]
Some parts of the code have wrong indentation (like in the `main` function) by the way.

File: dynamic_programming/longest_common_subsequence.cpp (hunk 22 above)
Comment (Santoshjonnakuti) [reply dynamic_programming/longest_common_subsequence.cpp]: Updated!

File: dynamic_programming/longest_common_subsequence.cpp [hunk 23]
```diff
@@ -0,0 +1,303 @@
...
+ * So Length of LCS will be 2.
+ *
+ * allPossibleLSC() will print all possible LCS between those strings.
+ *
+ * AC
+ * GC
+ * GA
+ * GA
+ * 
+ * To test run './longest_common_subsequence -test'
```
Comment (Panquesito7) [inline dynamic_programming/longest_common_subsequence.cpp:26]:
```suggestion
 * To test, run './longest_common_subsequence -test'
```

File: dynamic_programming/longest_common_subsequence.cpp [hunk 24]
```diff
@@ -0,0 +1,303 @@
...
+ * So There are 4 Longest common subsequences in 'GAC' and 'AGCAT'
```
Comment (Panquesito7) [inline dynamic_programming/longest_common_subsequence.cpp:27]:
```suggestion
 * So there are 4 longest common subsequences in `GAC`, and `AGCAT`
```

File: dynamic_programming/longest_common_subsequence.cpp [hunk 25]
```diff
@@ -0,0 +1,303 @@
...
+ */
+
+
+#include <stdio.h>      /// for I/O operations
+#include <stdlib.h>     /// for dynamic memory allocation
+#include <string.h>     /// for `puts()`, `gets()`, and `strlen()`
+#include <assert.h>     /// for `assert()`
+
+#define MAX 100      // maximum size of string1 and string2
```
Comment (Panquesito7) [inline dynamic_programming/longest_common_subsequence.cpp:36]:
```suggestion
#define MAX 100      ///< maximum size of string1 and string2
```

File: dynamic_programming/longest_common_subsequence.cpp [hunk 26]
```diff
@@ -0,0 +1,303 @@
...
+
+/**
+ * @brief Structure to store all LCS and no of possible LCS
+ */
+struct array
+{
+	char **LCSArray;      // array to store all possible LCS
+	int length;           // no of possible LCs
```
Comment (Panquesito7) [inline dynamic_programming/longest_common_subsequence.cpp:44]:
```suggestion
	char **LCSArray;      ///< array to store all possible LCS
	int length;           ///< no of possible LCs
```

File: dynamic_programming/longest_common_subsequence.cpp [hunk 27]
```diff
@@ -0,0 +1,303 @@
...
+		}
+	}
+	variable.LCSArray = LCSArray;
+	return;
+}
+
+/**
+ * @brief max will find the maximum between 2 values
+ * @param a integer a
+ * @param b integer b
```
Comment (Panquesito7) [inline dynamic_programming/longest_common_subsequence.cpp:128]: Please provide a better description of these parameters.

File: dynamic_programming/longest_common_subsequence.cpp [hunk 28]
```diff
@@ -0,0 +1,303 @@
...
+	{
+		return a;
+	}
+	// else return b
+	return b;
+}
+
+/**
+ * @brief LCS will make 2D-array with optimal solutions of sub-problems
+ * @param string1 String 1
```
Comment (Panquesito7) [inline dynamic_programming/longest_common_subsequence.cpp:144]: Please provide a better description.

File: dynamic_programming/longest_common_subsequence.cpp [hunk 29]
```diff
@@ -0,0 +1,303 @@
...
+ * @param m Length of String 1
+ * @param string2 String 2
```
Comment (Panquesito7) [inline dynamic_programming/longest_common_subsequence.cpp:146]: Please provide a better description.

File: dynamic_programming/longest_common_subsequence.cpp [hunk 30]
```diff
@@ -0,0 +1,303 @@
...
+			}
+		}
+	}
+	return c;
+}
+
+/**
+ * @brief findAllLCS is to print all possible LCS
+ * @param string1 string 1
+ * @param string2 string 2
```
Comment (Panquesito7) [inline dynamic_programming/longest_common_subsequence.cpp:197]: Please provide a better description of these parameters.

File: dynamic_programming/longest_common_subsequence.cpp [hunk 31]
```diff
@@ -0,0 +1,303 @@
...
+	int length;           // no of possible LCs
+} variable;
+
+
+// Functions Declaration
+int max(int a, int b);
+int** LCS(char *string1, int m, char *string2, int n);
+void allPossibleLCS(int **c, char *a, int index, int i, int j, char *string1, char *string2, int lengthOfLCS);
+void findAllLCS(char *string1, char *string2);
+static void test();
```
Comment (Panquesito7) [inline dynamic_programming/longest_common_subsequence.cpp:53]: I don't think this is needed. It could cause a bit more confusion to the reader IMO. Although it's not required to remove them, it's better to simply declare the function without its prototype.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: dynamic_programming/longest_common_subsequence.cpp [hunk 32]
```diff
@@ -0,0 +1,295 @@
...
+	return;
+}
+
+/**
+ * @brief max will find the maximum between 2 values
+ * @param a length of lcs between string1[m-1] and string2[n] m and n are lengths of strings
+ * @param b length of lcs between string1[m] and string2[n-1] m and n are lengths of strings
+ * @returns maximum value
+ */
+int max(int a, int b)
```
Comment (Panquesito7) [inline dynamic_programming/longest_common_subsequence.cpp:123]: C++ already includes a built-in `std::max` function. Please use that instead.
